enterprise/coordinator/recovery/recovery.go

```diff
@@ -0,0 +1,180 @@
+		return nil
+	}
+	r.logger.Info("Stale state observed, attempting recovery")
+	peers, err := r.peerGetter.GetPeers(ctx)
+	if err != nil {
+		return fmt.Errorf("getting peers: %w", err)
+	}
+	if len(peers) == 0 {
+		r.logger.Info("No peers to recover from")
+		return errNoPeers
+	}
+	var errs []error
+	for _, peer := range peers {
+		err := r.recoverFromPeer(ctx, oldState, net.JoinHostPort(peer, "7777"))
+		if err == nil {
+			return nil
+		}
+		r.logger.Warn("recovery attempt failed", "peer", peer, "err", err)
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+// recoverFromPeer sends a recovery request to the peer coordinator and recovers the state.
+func (r *Recoverer) recoverFromPeer(ctx context.Context, oldState *stateguard.State, peer string) error {
+	r.logger.Info("attempting recovery", "peer", peer)
+	_, err := r.guard.ResetState(oldState, authorizerFunc(func(mnfst *manifest.Manifest) (*seedengine.SeedEngine, *ecdsa.PrivateKey, error) {
+		validators, err := sdk.ValidatorsFromManifest(r.httpsGetter, mnfst, r.logger)
+		if err != nil {
+			return nil, nil, fmt.Errorf("generating validators: %w", err)
+		}
+
+		client, closeConn, err := r.dial(ctx, r.issuer, validators, r.logger, peer)
+		if err != nil {
+			return nil, nil, fmt.Errorf("dialing coordinator: %w", err)
+		}
+		defer func() {
+			if err := closeConn(); err != nil {
+				r.logger.Warn("Could not close connection", "err", err)
+			}
+		}()
+
+		resp, err := client.Recover(ctx, &meshapi.RecoverRequest{})
+		if err != nil {
+			return nil, nil, fmt.Errorf("calling Recover: %w", err)
+		}
+
+		se, err := seedengine.New(resp.Seed, resp.Salt)
+		if err != nil {
+			return nil, nil, fmt.Errorf("creating seed engine: %w", err)
+		}
+
+		block, _ := pem.Decode(resp.MeshCAKey)
+		meshCAKey, err := x509.ParseECPrivateKey(block.Bytes)
+		if err != nil {
+			return nil, nil, fmt.Errorf("parsing mesh CA key: %w", err)
+		}
+
+		return se, meshCAKey, nil
+	}))
+	if err != nil {
+		return fmt.Errorf("resetting state: %w", err)
+	}
+	return nil
+}
+
+func periodically(ctx context.Context, clock clock.WithTicker, interval time.Duration, f func()) error {
+	t := clock.NewTicker(interval)
+	defer t.Stop()
+	for {
+		f()
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C():
+		}
+	}
+}
+
+type authorizerFunc func(mnfst *manifest.Manifest) (*seedengine.SeedEngine, *ecdsa.PrivateKey, error)
+
+func (f authorizerFunc) AuthorizeByManifest(mnfst *manifest.Manifest) (*seedengine.SeedEngine, *ecdsa.PrivateKey, error) {
+	return f(mnfst)
+}
+
+func dial(ctx context.Context, issuer atls.Issuer, validators []atls.Validator, logger *slog.Logger, addr string) (meshapi.MeshAPIClient, func() error, error) {
+	dial := dialer.New(issuer, validators, atls.NoMetrics, nil, logger)
+	conn, err := dial.Dial(ctx, addr)
+	if err != nil {
+		return nil, nil, fmt.Errorf("dialing coordinator: %w", err)
+	}
+
+	client := meshapi.NewMeshAPIClient(conn)
+	return client, conn.Close, nil
+}
```

Comment:
I don't get why these have to be funcs, can't we just use a struct/method here to fulfill the interfaces?

Reply:
1. The dialer is a struct now.
2. The authorizer used to be a func so that I can capture all of the relevant arguments that only exist in that call. However, it turns out that I can just make the Recoverer itself an authorizer, I only need to pass a context and add the peer.